keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md

```diff
@@ -117,113 +129,79 @@ serves the resourceVersion="0" list requests from reflectors today.
-kubernetes "bookmark" watch event and send it to kube-apiserver clients so that
-reflectors and shared informers are kept up-to-date. The benefit of this is that
-it minimizes the chance that these clients will end up with an out-of-date
-resource version and need to relist (which can impact scalability). See [Watch
-Bookmarks](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190206-watch-bookmark.md)
-for details.
+### Risks and Mitigations
 
-#### Determining if etcd is sending progress notify events
+### Dependency on manual watch progress notifications
 
-It is possible to automatically determine if etcd is sending progress notify events.
-Each watch cache could keep track of when it received the last progress notify event.
-If it has been sufficiently long since the last one was received, or if none have
-ever been received, the watch cache should assume etcd is not sending progress notify,
-and not attempt to serve consistent reads from cache, falling back to serving
-the reads directly from etcd.
+Progress notify is requested on client, so all watchers opened with this client
+will get the notification. This is not a problem for Kubernetes as we maintain
+one client per resource type, but could cause an issue if we attempt to reuse
+the client across resource types (kubernetes/kubernetes#114458). This issue can
+be mitigated by having multiple grpc watch streams within a single etcd client.
 
-### Risks and Mitigations
+When etcd opens a watch it assigns it a stream based on its [context metadata].
+So by default all watches opened on single client share single grpc stream.
+To implement reusing etcd for multiple resources we should consider adding
+unique metadata for each resource, forcing etcd client to create a separate
+grpc for each of them. This way requesting progress notification on a specific
+resource will result in only that single watch being notified.
 
-Configuring etcd to send progress notify events may have performance implications.
-@mm4tt and @wojtek-t have run [some
-experiments](https://github.com/kubernetes/kubernetes/pull/86769) that suggest
-that a progress notify interval of 50ms results in a noticeable increase in both
-etcd and kube-apiserver CPU utilization and some increased kube-apiserver serving
-latency. We intend to address this by first trying a 250ms progress notify interval,
-if ths performs well we will use that as our interval. If not, we will dig in
-to figure out why the events are impacting performance and see if we can optimize
-it away.
+[context metadata]: https://github.com/etcd-io/etcd/blob/a6ab774458411a6c0ea08f5df97e4dcc9a836345/client/v3/watch.go#L1070-L1075
 
-## Design Details
+<<[UNRESOLVED @serathius]>>
+Propose how kube-apiserver should fallback if etcd watch notification doesn't work.
+Progress notification was introduced v3.4 etcd release but was affected by bug
+etcd-io/etcd#15220 that was fixed in v3.4.25 and v3.5.8.
```

Comment:
My concern is that neither of the options below really protect us from the bug version:
- (1) - minor version 3.4 is not enough (e.g. 3.4.21 is buggy)
- (2) - one can enable it on a bad etcd version
- (3) - maybe this one works, but it's somewhat ugly...
- (4) - what if progress brings a bad value which the bug was about

But I think it's fine to make an Alpha without it (and for Alpha rely purely on documentation).

Reply:
For alpha, I agree, tracking this with an UNRESOLVED (for the KEP) and with documentation (for the user) is enough.

Reply:
Added note about documenting this.